flips/20210916-nft-metadata.md

```diff
@@ -0,0 +1,354 @@
+protocol: http
+
+An on-chain image:
+data: data:image/png;base64,SOMEPNGDATAURI/wD/
+contentType: image/jpeg
+protocol: data
+
+The problem here is that the on-chain image can be quite large and you might not want to resolve those every time you fetch the media view. So what if we made a method on the struct that allowed you to retrieve the data/content instead?
+
+
+The same NFT author may also want expose other metadata about the artwork:
+```
+struct Artwork {
+    let name: String
+    let dateCreated: UInt64
+    let medium: String
+    let artistName: String
+}
+
+// ...
+
+let artwork = nft.resolveView(Type<ArtNFT.Artwork>()) as? ArtNFT.Artwork
+```
+
+If an art piece had more than one image then they could be retrieved using an array type as such:
+```
+import ArtNFT from 0x01010101
+   
+// ...
+   
+let images = nft.resolveView(Type<[ArtNFT.IPFSImage]>()) as? [ArtNFT.IPFSImage]
+```
+
+Or if a more structured approach is desired the NFT author may want to create a composite struct:
+```
+struct ArtworkDetail {
+    let artwork: Artwork
+    let primaryImage: IPFSImage
+    let secondaryImage: IPFSImage
+}
+```
+
+Because this proposal relies on the Cadence type system, adding new view types for exposing various
+types of metadata is trivial and very robust. A few examples that one might use as a view type are:
+
+```
+// an image
+pub struct IPFSImage {
+    pub let url: String
+    pub let ipfsHash: String
+    pub let width: UInt32
+    pub let height: UInt32
+    pub let contentType: String
+}
+
+// a video
+pub struct IPFSVideo {
+    pub let url: String
+    pub let ipfsHash: String
+    pub let width: UInt32
+    pub let height: UInt32
+    pub let codec: String
+    pub let container: String
+    pub let contentType: String
+}
+
+// a file stored on the blockchain
+pub struct Mime {
+    pub let type: String
+    pub let bytes: [UInt8]
+}
+
+// a location on the planet
+pub struct GeoPoint {
+    pub let lat: UFix64
+    pub let lng: UFix64
+    pub let projection: String
+}
+
+// a file hosted by a web server with a hash for verifying it's contents
+pub struct URLWithHash {
+    pub let url: String
+    pub let hash: String?
+    pub let hashAlgo: String?
+}
+```
+
+### Royalty 
+
+This particular problem has been discussed extensivly. The following suggestion was agreed upon
+by @bjartek, @rheaplex and @dete. 
+
+```
+  /// a struct interface that describes how to calculate royalty 
+	pub struct interface Royalty {
+
+		/// if nil cannot pay this type
+		/// if not nill withdraw that from main vault and put it into distributeRoyalty 
+		pub fun calculateRoyalty(type:Type, amount:UFix64) : UFix64?
+
+		/// call this with a vault containing the amount given in calculate royalty and it will be distributed accordingly
+		pub fun distributeRoyalty(vault: @FungibleToken.Vault) 
+
+		/// generate a string that represents all the royalties this NFT has for display purposes
+		pub fun displayRoyalty() : String?  
+
+	}
+```	
```

Comment:
I think this is little problematic to have `calculateRoyalty` and `distributeRoyalty` here, wouldn't it let to contract owner to update contract and then get all sale revenue to himself?  

or can change artist royalty address etc ? 

So far we addressed things to be usually read-only, but this is a big leap

Reply:
This suggestion was made by Dieter. So I think adding this to the relevant place in the NFT repo for discussing views is the right thing to do.